ref(insights): Improve async dropdown load behaviour

gggritso opened the pull request:
We have a few UI dropdowns that fetch more data on search. The Queries page "Table" selector and the "Assets" page "Page" selector are two prominent examples. When a user types into that box, we make a server request to fetch possible matches, since there are more items than we can load on page load.

The two dropdowns have different behaviour, which is yucky. This PR consolidates them.

1. Moves `TransactionSelector` into the `common` directory, out of its parent, so it can be re-used
2. Extracts the caching logic for dropdown results

`TransactionSelector` and `DomainSelector` code looks very similar now, which is good news! The _main_ benefit though, is they both have the same behaviour:

- the dropdown contents are _always available_ even if we're currently loading more possible matches
- a loading spinner shows when data is loading, instead of loading text
- we don't make extra requests if there are < 100 possible total matches and we've loaded them all
- the current selection from URL is always shown, even if it wasn't in the autocomplete fetch results

**Before:**

https://github.com/user-attachments/assets/98aff86e-858a-46bf-92ac-faa3c7c20965


**After:**

https://github.com/user-attachments/assets/0405ae74-ce1c-4c81-a998-19c281469400


This is a pre-requisite for replacing both of these with a `CompactSelect`, which is what I'll be doing soon. There's room to combine these new hooks somehow, but I'm not worried about that yet.

gggritso committed:
Add an array cache hook

gggritso committed:
Use array cache hook in `DomainSelector`

gggritso committed:
Extract cache check behaviour into a hook

gggritso committed:
Move transaction selector into its own file

gggritso committed:
Align `DomainSelector` and `TransactionSelector`

gggritso committed:
Add documentation string

gggritso committed:
Simplify implementation

- remove `mergeFn`, we don't need it
- use `useMemo` for a simpler interfact

gggritso committed:
Add spec

gggritso committed:
Clarify comment

gggritso committed:
Clarify comment

gggritso committed:
Bring back `mergeFn`

gggritso committed:
Preserve page value

gggritso committed:
Fix double value bug

JonasBa commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+  return items;
+}
+
+export function setMerge<T>(items: T[]): T[] {
```
Unless this is a default merge strategy, I would let callers define it

gggritso replied:
😭 I know, but this one is _so common_ I wonder where I can put this

JonasBa replied:
Just inline it at each spot :D it's short enough hehe

gggritso commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+ * Every time the `useArrayCache` hook is called with new items it adds
+ * them to the cache, and returns all known items.
+ */
+export function useArrayCache<T>(props: Props<T>): T[] {
```
> Isnt this just a list of sorts with a utility to merge and sort two lists?

No? It's a cache that keeps a record of _all items that have ever been passed to it_, not just two lists.

> I find it weird when we name things by their backing data structure

That's fair, but generic utilities need generic names. This is a cache of a data structure that's both sortable and mergeable, so I went with `arrayCache`

JonasBa replied:
> No? It's a cache that keeps a record of all items that have ever been passed to it, not just two lists.

I'm struggling to wrap my mind about this tbh. Seeing that mergeSet is the common merge strategy makes me question the use of arrays. Maybe arrays are what you want as the output, but using it to back a cache implementation feels weird

JonasBa replied:
>That's fair, but generic utilities need generic names. This is a cache of a data structure that's both sortable and mergeable, so I went with arrayCache

Tbh, I dont have a better name for it. Maybe something like argumentHistoryCache or something. Maybe worth getting a few eyes on this see if someone has a good idea how to name something like this

JonasBa replied:
On a second note, how about using an LRU cache and sorting the values before returning from the hook?

gggritso commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+  const {items: incomingItems, sortFn, mergeFn, limit = -1} = props;
+
+  // The cache is a reference. This makes it possible to update the cache without causing a re-render, which avoids an infinite render loop.
+  const cache = useRef<Cache<T>>({
```
We're not avoiding re-renders, we're avoiding an infinite re-render loop! We definitely want re-renders when the cache contents change

gggritso commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    if (mergeFn) {
+      newItems = mergeFn(newItems);
```
Yeah, I think the "mergeFn" here is probably not a good name, something else would be better. Let me think about it! As for merge vs. sort first, I was torn, but in a lot of cases you want to sort first because merging sorted arrays is a lot faster, and it would be annoying to have to do that twice (once for sort, and then one more time for merge) but it's not a strong opinion on my part

Also, what is "gen" in this context?

JonasBa replied:
Oh sorry! I meant gen as the index generation that we are merging.

For sorting and merging, how about sorting each gen separately and merging them afterwards?

gggritso commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+ * Every time the `useArrayCache` hook is called with new items it adds
+ * them to the cache, and returns all known items.
+ */
+export function useArrayCache<T>(props: Props<T>): T[] {
```
Good one, yes! Will change

gggritso commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+ * Every time the `useArrayCache` hook is called with new items it adds
+ * them to the cache, and returns all known items.
+ */
+export function useArrayCache<T>(props: Props<T>): T[] {
```
👍🏻 for sure

gggritso commented on static/app/views/insights/common/utils/useArrayCache.tsx:
```diff
@@ -0,0 +1,69 @@
+import {useMemo, useRef} from 'react';
```
Sorry, what is "ds" in this context? I'm not sure what you mean overall, since I'm not avoiding renders or bypassing React, rather avoiding an infinite re-render loop

JonasBa replied:
Sorry lol, I meant the data structure that backs the insertion mechanism :)

gggritso commented:
@JonasBa I feel like maybe the issue here is that the way I factored this code doesn't make sense, and so it's confusing. I worked hard to make it very generic, and lost meaning as a result.

I'm thinking of maybe renaming this to `useCompactSelectOptionsCache` and making the behaviour more specific/meaningful and maybe adding some more features into it.

Thoughts?

JonasBa commented:
> @JonasBa I feel like maybe the issue here is that the way I factored this code doesn't make sense, and so it's confusing. I worked hard to make it very generic, and lost meaning as a result.

You nailed it! I failed to put it into words about what exactly didnt feel right, but I dont think you could have put it any better. Naming it `useCompactSelectOptionsCache` would clear up all the confusion and scope it to a use case which clears up the intent behind this code

gggritso committed:
Merge branch 'master' into ref/insights/dropdown-options-cache

gggritso committed:
Add `useCompactSelectOptionsCache` hook

gggritso committed:
Use `useCompactSelectOptionsCache`

Forget the array cache, it's fine.

gggritso committed:
Remove unused hook

gggritso commented:
@JonasBa I threw out the whole hook! I kept the new hook limited to just the functionality I actually need _right now_, so no need to futz with sorting or merging functions. I also used a `Map` as the cache, which I think is a better fit, and took some of the suggestions from your first review.

Give it a 👀

JonasBa commented on static/app/views/insights/common/utils/useCompactSelectOptionsCache.tsx:
```diff
@@ -0,0 +1,36 @@
+ * 1. Display an options dropdown even when new results are loading
+ * 2. Shows options from 3 searches ago if the user changes their mind
+ */
+export function useCompactSelectOptionsCache(options: Option[]): readonly Option[] {
+  const cache = useRef<OptionCache>(new Map());
+
+  return useMemo(() => {
+    options.forEach(option => {
+      cache.current.set(option.value, option);
+    });
+
+    return Array.from(cache.current.values()).sort(alphabeticalCompare);
+  }, [options]);
+}
+
+type OptionComparator = (a: Option, b: Option) => number;
+
+const alphabeticalCompare: OptionComparator = (a, b) => {
+  return a.value.toString().localeCompare(b.value.toString());
+};
```
Love it, this is so much simpler!

gggritso committed:
Update API

Return an _object_ instead of just the plain array. This lets the hook
return more than just the computed options.

gggritso committed:
Return a cache clearing callback

gggritso committed:
Clear cache when project changes

gggritso committed:
Filter out empty values

DominikB2014 reviewed (approved):
Very nice 👍
